geom/interval.go

```diff
@@ -0,0 +1,28 @@
+package geom
+
+// Interval represents the interval bound by two float64 values. The interval
+// is closed, i.e. its endpoints are included. An interval typically has
+// distinct endpoints (i.e. is non-degenerate). It may also be degenerate and
+// contain no elements, or degenerate and contain a single element (i.e. the
+// min and max bounds are the same). The zero value of Interval is the
+// degenerate interval that contains no elements.
+type Interval struct {
+	min, max float64
+	nonEmpty bool
+}
+
+// NewInterval returns a new non-empty Interval with the given bounds (which
+// may be the same).
+func NewInterval(boundA, boundB float64) Interval {
+	if boundB < boundA {
+		boundA, boundB = boundB, boundA
+	}
+	return Interval{boundA, boundB, true}
+}
+
+// MinMax returns the minimum and maximum bounds of the interval. The boolean
+// return value indicates if the interval is non-empty (the minimum and maximum
+// bounds should be ignored if false).
+func (i Interval) MinMax() (float64, float64, bool) {
```

Comment:
Another idiom I've seen used is with `time.Time` and `uuid.UUID` (google's) which use `IsZero` and `uuid.Nil` respectively. I wonder if we should follow the same idiom?

Reply:
This is something I considered, but I don't think it's the way to go for the following reasons. I'll use `time.Time` as the point of comparison:

- `time.Time`'s zero value represents the instant "Jan 1, year 1, 00:00:00 UTC", and is detected by `IsZero()`.
- That's a valid time, albeit one that would be unlikely to need to be represented in practice.
- Mapping that over to `Interval`, the zero value would be the interval `[0, 0]` (i.e. the interval containing just the number `0`).
- But the `[0, 0]` interval would be useful in _a lot_ of contexts that aren't rare in the same way that "Jan 1, year 1, midnight" is rare. For example, as an `ExtendedEnvelope`, the interval might indicate a range of elevations that is perfectly flat (i.e. all zero height compared to some reference height). That needs to be differentiated from an `ExtendedEnvelope` that doesn't have any elevation data.
- The other approach could be to keep the `nonEmpty` boolean inside `Interval`, but change the signature of `MinMax` to `MinMax() (float64, float64)` and add an `IsZero() bool` method (which might have been what you were thinking?). If `IsZero()` returns `true`, then the return value of `MinMax()` should be ignored. That would represent everything ok, but I'm really not a fan of those types of APIs, since it's very easy for users to miss that the type represents a potentially empty interval (i.e. they just see the `MinMax` method and ignore the `IsZero` method, not realising that `Interval` may represent an empty interval).

Reply:
Good points, and agree!